### PR TITLE
Fixes #196 - htmlbool is now false when 'false' or '0'

### LINKED ISSUE #196: 'htmlbool' type is never false when set to a "false" or "0"
__How often can you reproduce it?__ 

- [X] Always
- [ ] Sometimes
- [ ] Rarely
- [ ] Unable
- [ ] I didn’t try

__Description:__

Setting an 'htmlbool' attribute on a can-component is never set to `false` except in the case of the attribute not existing on the element.

For example:
`<foo-component expanded /> //-> expanded === '' === true` (per the documentation)
`<foo-component expanded="false" /> //-> expanded === "false" === true`

__Steps to reproduce:__

1. Create a component with an html attribute equal to false `<foo-component expanded="false" />`
2. On the above component, add a viewModel property `expanded` with a `type: 'htmlbool'`
3. Put something in your template that relies on that property being true/false, (i.e. `{{#if expanded}} .... {{/if}}`)

__Expected results:__
Per the documentation (https://canjs.com/doc/can-define.types.html), `htmlbool` should act like boolean ("false", "0", should be false) except for the fact that empty string is `true`.

Therefore `<foo-component expanded="false" />`, the viewModel property `expanded` should be falsey.

__Actual results:__

The viewModel property `expanded` (from the example above) is truthy.

__Analysis:__
Either the documentation is wrong, or we have a bug at:
https://github.com/canjs/can-define/blob/master/can-define.js#L862

The `typeof val === "string"` will short circuit when `expanded="false"`. 

I believe the `htmlbool` should be changed to:

```js
if (val === '') return true;
return define.types.boolean.apply(this, val);
``` 

__Environment:__

| Software                   | Version
| -------------------- | -------
| can-define version   | 1.0.17

### PR DESCRIPTION
**Pull request by mjstahl**
Fixes #196  - 'htmlbool' type is never false when set to a "false" or "0"

### CODE REVIEW

File: can-define.js [hunk 1]
```diff
@@ -859,7 +859,8 @@ define.types = {
 	 * any string, including "", is truthy.
 	 */
 	'htmlbool': function(val) {
-		return typeof val === "string" || !!val;
+		if (val === '') return true;
```
Comment (justinbmeyer) [inline can-define.js]:
```
if (val === '') {
  return true;
}
```

Comment (mjstahl) [inline can-define.js]: Yeah, the linter yelled at me. Good catch though.

File: can-define.js [hunk 2]
```diff
@@ -859,7 +859,13 @@ define.types = {
 	 * any string, including "", is truthy.
 	 */
 	'htmlbool': function(val) {
-		return typeof val === "string" || !!val;
+		if (val === '') {
+			return true;
+		}
+		if (val == null) {
```
Comment (mjstahl) [inline can-define.js]: This null/undefined check is ensure no participation in https://github.com/canjs/canjs/issues/2316 (null/undefined should bypass type coercion).

**Comment by mjstahl** [comment]
Though this PR ensure we are compliant with the documentation, it breaks current functionality.

In reality, I believe setting an `htmlbool` property to any falsey value should actually remove the attribute from the element, but I am not sure how that would hold up with binding.

I believe this to be a good compromise and it puts the code in line with our documentation.

File: can-define.js (hunk 2 above)
Comment (justinbmeyer) [inline can-define.js]: ensure no participation?  Do you mean comply with 2316?  I thought this was an exception.

This is for scenarios like:

```js
VM = DefineMap.extend({
  prop: "htmlbool"
})
```

```
<component {{#something}}prop{{/something}}>
```

In this case, `prop` will be set to `""` so need to be `true`.

I'm not sure what the following should do:

```
<component {{#something}}prop="false"{{/something}}>
```

perhaps setting to false is right.

What you might want to use is https://canjs.com/doc/can-util/js/string-to-any/string-to-any.html



```
!!stringToAny(value)
```

Comment (mjstahl) [inline can-define.js]: Good call. The PR is updated.

**Pull request merged**